spec/src/main/asciidoc/ch04-query-language.adoc

```diff
@@ -681,18 +705,23 @@ Example:
 SELECT c FROM Customer c, Employee e WHERE c.hatsize = e.shoesize
 ----
 
-In general, use of this style of inner join
-(also referred to as theta-join) is less typical than explicitly defined
-joins over relationships.
+This style of inner join (sometimes called a "theta" join) is less
+typical than explicitly defined joins over relationships.
 
-The syntax for explicit join operations is as
-follows:
+The syntax for explicit join operations is given by:
 
 ----
-join ::= join_spec join_association_path_expression [AS] identification_variable [join_condition]
+join ::= range_join | path_join | fetch_join
+
+range_join ::= join_spec range_variable_declaration [join_condition]
+
+path_join ::=
+    join_spec join_association_path_expression [AS] identification_variable [join_condition]
 
 fetch_join ::= join_spec FETCH join_association_path_expression
 
+join_spec ::= [INNER | LEFT [OUTER] | RIGHT [OUTER] | FULL [OUTER]] JOIN
```

Comment:
The new join types should also be added to the `jakarta.persistence.criteria.JoinType` enum.
While we're at it, we should IMO also add `cross join` so that people don't have to write `a join b on 1=1`.

We might have to adapt some texts that talk about portability of right joins e.g.

> Right outer joins and right outer fetch joins are not required
> to be supported. Applications that use <code>RIGHT</code> join
> types will not be portable.

Either we should add the FULL join type to this portability note or "specify" the behavior to make it a required feature.

Reply:
> We might have to adapt some texts that talk about portability of right joins

OK, good point.

Reply:
> The new join types should also be added to the `jakarta.persistence.criteria.JoinType` enum.

Interestingly, `JoinType` already has `RIGHT`, even though the spec never mentions right joins anywhere. But yes, need to add `FULL`.

Reply:
> we should IMO also add `cross join` so that people don't have to write `A a join B b on 1=1`

Why can't they just just write `A a, B b`?

I guess I'm not convinced we need to add `cross join`.

Reply:
>Why can't they just just write `A a, B b`?

If you want to be able to refer to `a` in a following join, then you must use a join between `a` and `b`, because the comma separator isolates from clause spaces i.e. it is illegal to write e.g. `from A a, B b join C c on b.x = c.x and a.y = c.y` because `a` is not visible within the second from clause space. Right now, people have to write `from A a join B b on 1=1 join C c on b.x = c.x and a.y = c.y` but I think it might be nicer to allow writing the cross join instead `from A a cross join B b join C c on b.x = c.x and a.y = c.y`.

Reply:
> If you want to be able to refer to a in a following join, then you must use a join between a and b, because the comma separator isolates from clause spaces

Ummmm, which section of the JPA spec says that?

Reply:
> AFAIU, this means the comma separator must be rendered to a cross join in SQL

Well, you only need to do this transformation in the case where there's a reference to the identification variable before the comma from a join occurring after the comma. Which is not going to be the common case, and seems like it's easy to detect.

Reply:
Well, "easy" might be a slight overstatement. But _possible_, surely.

Reply:
A quick test with PostgreSQL showed that the comma separator vs. cross join does indeed affect the join order and hence the execution plan might differ when 3+ tables are involved. Here is a nice overview showing the differences: https://stackoverflow.com/a/54738895/412446

It seems very hard (impossible?) to express the following query with the Criteria API

```
from A a cross join B b join C c on ...
```

e.g.

```
CriteriaQuery<?> cq = ...
Root<A> a = cq.from(A.class);
Root<B> b = cq.from(B.class);
Join<C> c = a.join(C.class);
```

unless the `CriteriaQuery` is supposed to have a fixed order of from clause items based on the order in which methods are called. In Hibernate, we currently only order joins within a `Root`.

I'm trying to understand if the cross join support we have in Hibernate is "useless" and that we should implement join ordering for the whole query instead of per `Root`, or if the JPA spec might need cross joins as well.
Right now, it seems to me that we should get rid of cross joins in Hibernate, but I don't yet understand if the join ordering implications of rendering a SQL cross join might have a negative effect on performance. Either way, the spec seems to imply that the SQL must have cross joins semantics, so it seems like a forced move.